### PR TITLE
Feature/optimization

### PR DESCRIPTION
**Pull request by afeena**
Some improvements to make TANNER faster

### CODE REVIEW

File: server.py [hunk 1]
```diff
@@ -5,13 +5,18 @@
...
 import session_manager
 import api
 import base_handler
 import logger
+import uvloop
 
+asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
 logger = logger.Logger.create_logger('tanner.log', 'tanner')
+redis_client = asyncio.get_event_loop().run_until_complete(
+    asyncio_redis.Pool.create(host='localhost', port=6379, poolsize=10))
```
Comment (afeena) [inline server.py]: It was the most tricky question for me - how to organize new asyncio redis. Before I initialize redis for every emulator in init, but I can't use generators in __init__. This works, but maybe there is more elegant (and right) way?

File: dorks_manager.py [hunk 2]
```diff
@@ -15,13 +16,14 @@ class DorksManager:
...
-        if not self.r.exists(self.dorks_key):
-            self.push_init_dorks('dorks.pickle', self.dorks_key)
-        if not self.r.exists(self.user_dorks_key):
-            self.push_init_dorks('user_dorks.pickle', self.user_dorks_key)
+    def __init__(self, redis_client):
+        self.redis = redis_client
+        # if not (yield from self.redis.exists(self.dorks_key)):
+        #  self.push_init_dorks('dorks.pickle', self.dorks_key)
+        #   if not (yield from self.redis.exists(self.user_dorks_key)):
+        # self.push_init_dorks('user_dorks.pickle', self.user_dorks_key)
```
Comment (glaslos) [inline dorks_manager.py]: Remove unused code

File: dorks_manager.py [hunk 3]
```diff
@@ -40,13 +43,16 @@ def extract_path(self, path):
...
 
+    @asyncio.coroutine
     def choose_dorks(self):
         chosen_dorks = []
         max_dorks = 50
-        dorks = self.r.smembers(self.dorks_key)
-        user_dorks = self.r.smembers(self.user_dorks_key)
+        dorks = yield from self.redis.smembers(self.dorks_key)
+        dorks = yield from dorks.asset()
+        user_dorks = self.redis.smembers(self.user_dorks_key)
```
Comment (glaslos) [inline dorks_manager.py]: no yield?

File: lfi_emulator.py [hunk 4]
```diff
@@ -37,15 +41,19 @@ def setup_vdocs(self):
             if not files:
                 with open('data/vdocs.json') as vdf:
                     vdocs = json.load(vdf)
+                vdf.close()
```
Comment (glaslos) [inline lfi_emulator.py]: not required, leaving the `with` context closes the file

File: lfi_emulator.py [hunk 5]
```diff
@@ -37,15 +41,19 @@ def setup_vdocs(self):
...
         if vdocs:
             for k, v in vdocs.items():
                 filename = self.vdoc_path + 'linux/' + k
                 os.makedirs(os.path.dirname(filename), exist_ok=True)
                 with open(filename, 'w') as vd:
                     vd.write(v)
+                vd.close()
```
Comment (glaslos) [inline lfi_emulator.py]: not required, leaving the with context closes the file

File: rfi_emulator.py [hunk 6]
```diff
@@ -33,30 +33,34 @@ def download_file(self, path):
         except aiohttp.ClientError as e:
             self.logger.error('Error during downloading the rfi script'.format(e))
         else:
-            resp.release()
+            yield from resp.release()
+            yield from client.close()
             file_name = hashlib.md5(data.encode('utf-8')).hexdigest()
             with open(self.script_dir + file_name, 'w') as rfile:
                 rfile.write(data)
+            rfile.close()
```
Comment (glaslos) [inline rfi_emulator.py]: not required, leaving the with context closes the file

File: rfi_emulator.py [hunk 7]
```diff
@@ -33,30 +33,34 @@ def download_file(self, path):
...
     def get_rfi_result(self, path):
-        yield from asyncio.sleep(1)
         rfi_result = None
+        yield from asyncio.sleep(1)
         file_name = yield from self.download_file(path)
         if file_name is None:
             return rfi_result
         with open(self.script_dir + file_name) as f:
             script_data = f.read()
+        f.close()
```
Comment (glaslos) [inline rfi_emulator.py]: not required, leaving the with context closes the file

File: server.py (hunk 1 above)
Comment (glaslos) [inline server.py]: Have you tried:
```Python
redis_client = None

def get_redis_client():
    if not redis_client:
        redis_client = client()
    return redis_client
```
My name scoping might be off, writing more Go lately.

**Comment by glaslos** [comment]
Something is wrong with the test...

**Comment by afeena** [comment]
The problem is in the ``redis_client`` init, Travis can't connect to redis, but it trying to connect (because it's a static variable which I can mock only before import), and it is the infinite loop in the server. I'm trying to fix it

File: sqli_emulator.py [hunk 8]
```diff
@@ -66,13 +66,13 @@ def create_attacker_db(self, session):
...
     @staticmethod
     def prepare_get_query(path):
-        path = urllib.parse.unquote(path)
         query = urllib.parse.urlparse(path).query
         parsed_query = urllib.parse.parse_qsl(query)
         return parsed_query
 
     @asyncio.coroutine
     def map_query(self, query):
+        print(self.query_map)
```
Comment (glaslos) [inline sqli_emulator.py]: remove this

File: api.py [hunk 9]
```diff
@@ -8,36 +8,37 @@
 class Api:
     def __init__(self):
         self.logger = logging.getLogger('tanner.api.Api')
-        self.r = redis.StrictRedis(host='localhost', port=6379, decode_responses=True)
+        # self.redis = redis_client
```
Comment (glaslos) [inline api.py]: remove this

**Comment by glaslos** [comment]
Looks good! Just two minor fixes. Yield all the things 😄

**Comment by afeena** [comment]
Thanks :)

**Pull request merged**